### ISSUE TITLE
Hard crash due to "No such attribute testonly in new_http_archive rule //external:zlib"

### ISSUE THREAD

**Issue by endobson** [bazelbuild/bazel#3676]
Source code:
https://github.com/endobson/examples/tree/testonly-crash

bazel version 0.5.4

```
endobson@yggdrasil () ~/proj/bes-example % bazel build @googleapis//:build_event_service_grpc_cc                                                                                                                                          (0)
INFO: Reading 'startup' options from /Users/endobson/.bazelrc: --watchfs
.............
Unhandled exception thrown during build; message: Unrecoverable error while evaluating node 'CONFIGURED_TARGET:@grpc//:grpc_base_c 1bde2b824d8d0d8b964fcc970e455a65 (252815988 848607884)' (requested by nodes 'CONFIGURED_TARGET:@grpc//:grpc_base 1bde2b824d8d0d8b964fcc970e455a65 (1365554887 848607884)')
INFO: Elapsed time: 4.727s
java.lang.RuntimeException: Unrecoverable error while evaluating node 'CONFIGURED_TARGET:@grpc//:grpc_base_c 1bde2b824d8d0d8b964fcc970e455a65 (252815988 848607884)' (requested by nodes 'CONFIGURED_TARGET:@grpc//:grpc_base 1bde2b824d8d0d8b964fcc970e455a65 (1365554887 848607884)')
        at com.google.devtools.build.skyframe.ParallelEvaluator$Evaluate.run(ParallelEvaluator.java:475)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:352)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: No such attribute testonly in new_http_archive rule //external:zlib
        at com.google.devtools.build.lib.packages.AbstractAttributeMapper.getIndexWithTypeCheck(AbstractAttributeMapper.java:229)
        at com.google.devtools.build.lib.packages.AbstractAttributeMapper.get(AbstractAttributeMapper.java:57)
        at com.google.devtools.build.lib.packages.NonconfigurableAttributeMapper.get(NonconfigurableAttributeMapper.java:45)
        at com.google.devtools.build.lib.bazel.rules.BazelPrerequisiteValidator.isTestOnlyRule(BazelPrerequisiteValidator.java:130)
        at com.google.devtools.build.lib.bazel.rules.BazelPrerequisiteValidator.validateDirectPrerequisiteForTestOnly(BazelPrerequisiteValidator.java:113)
        at com.google.devtools.build.lib.bazel.rules.BazelPrerequisiteValidator.validate(BazelPrerequisiteValidator.java:38)
        at com.google.devtools.build.lib.analysis.RuleContext$Builder.validateDirectPrerequisite(RuleContext.java:2078)
        at com.google.devtools.build.lib.analysis.RuleContext$Builder.createTargetMap(RuleContext.java:1794)
        at com.google.devtools.build.lib.analysis.RuleContext$Builder.build(RuleContext.java:1621)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createRule(ConfiguredTargetFactory.java:315)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createConfiguredTarget(ConfiguredTargetFactory.java:228)
        at com.google.devtools.build.lib.skyframe.SkyframeBuildView.createConfiguredTarget(SkyframeBuildView.java:494)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.createConfiguredTarget(ConfiguredTargetFunction.java:1164)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:269)
        at com.google.devtools.build.skyframe.ParallelEvaluator$Evaluate.run(ParallelEvaluator.java:400)
        ... 4 more
java.lang.RuntimeException: Unrecoverable error while evaluating node 'CONFIGURED_TARGET:@grpc//:grpc_base_c 1bde2b824d8d0d8b964fcc970e455a65 (252815988 848607884)' (requested by nodes 'CONFIGURED_TARGET:@grpc//:grpc_base 1bde2b824d8d0d8b964fcc970e455a65 (1365554887 848607884)')
        at com.google.devtools.build.skyframe.ParallelEvaluator$Evaluate.run(ParallelEvaluator.java:475)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:352)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.IllegalArgumentException: No such attribute testonly in new_http_archive rule //external:zlib
        at com.google.devtools.build.lib.packages.AbstractAttributeMapper.getIndexWithTypeCheck(AbstractAttributeMapper.java:229)
        at com.google.devtools.build.lib.packages.AbstractAttributeMapper.get(AbstractAttributeMapper.java:57)
        at com.google.devtools.build.lib.packages.NonconfigurableAttributeMapper.get(NonconfigurableAttributeMapper.java:45)
        at com.google.devtools.build.lib.bazel.rules.BazelPrerequisiteValidator.isTestOnlyRule(BazelPrerequisiteValidator.java:130)
        at com.google.devtools.build.lib.bazel.rules.BazelPrerequisiteValidator.validateDirectPrerequisiteForTestOnly(BazelPrerequisiteValidator.java:113)
        at com.google.devtools.build.lib.bazel.rules.BazelPrerequisiteValidator.validate(BazelPrerequisiteValidator.java:38)
        at com.google.devtools.build.lib.analysis.RuleContext$Builder.validateDirectPrerequisite(RuleContext.java:2078)
        at com.google.devtools.build.lib.analysis.RuleContext$Builder.createTargetMap(RuleContext.java:1794)
        at com.google.devtools.build.lib.analysis.RuleContext$Builder.build(RuleContext.java:1621)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createRule(ConfiguredTargetFactory.java:315)
        at com.google.devtools.build.lib.analysis.ConfiguredTargetFactory.createConfiguredTarget(ConfiguredTargetFactory.java:228)
        at com.google.devtools.build.lib.skyframe.SkyframeBuildView.createConfiguredTarget(SkyframeBuildView.java:494)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.createConfiguredTarget(ConfiguredTargetFunction.java:1164)
        at com.google.devtools.build.lib.skyframe.ConfiguredTargetFunction.compute(ConfiguredTargetFunction.java:269)
        at com.google.devtools.build.skyframe.ParallelEvaluator$Evaluate.run(ParallelEvaluator.java:400)
        ... 4 more
```

**Comment by endobson** [comment]
This looks like it is partially caused by the new_http_archive and bind call in the WORKSPACE file both using the name "zlib". If I change them to be different then it doesn't crash.

**Comment by meteorcloudy** [comment]
Looks like we have the solution here, I'm closing this issue, please reopen if having any more questions.

**Comment by endobson** [comment]
I have a workaround not a solution, so please reopen (the acls don't let me do it myself.)

The bug here is that bazel crashes on some input, which it should never do, it should instead nicely print an error message if there is an error.

There is also the problem that the workspace file is accepted, and many things work but some particular configuration of rules don't like the reused names. Either using the name twice is an error and there should be always be an error message in that scenario or what ever thing is erroring out needs to figure out how to handle this situation and not produce an error.

**Comment by dslomov** [comment]
https://bazel-review.googlesource.com/c/bazel/+/31892 is the work towards fixing the hard crash. We need to revisit the general question of duplicate names (see also #588, #1952)

**Comment by dslomov** [comment]
Hard crash is fixed in 822a8b3
